libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt

```diff
@@ -730,34 +747,172 @@ class MapboxRouteLineApi(
         updateUpcomingRoutePointIndex(routeProgress)
         updateVanishingPointState(routeProgress.currentState)
 
+        val routeLineMaskingLayerDynamicData = when (
+            (routeProgress.currentLegProgress?.legIndex ?: INVALID_ACTIVE_LEG_INDEX) !=
+                activeLegIndex
+        ) {
+            true -> getRouteLineDynamicDataForMaskingLayers(currentPrimaryRoute, routeProgress)
+            false -> null
+        }
+        activeLegIndex = routeProgress.currentLegProgress?.legIndex ?: INVALID_ACTIVE_LEG_INDEX
+
         // If the de-emphasize inactive route legs feature is enabled and the vanishing route line
         // feature is enabled and the active leg index has changed, then calling the
         // alternativelyStyleSegmentsNotInLeg() method here will get the resulting calculation cached so
         // that calls to alternativelyStyleSegmentsNotInLeg() made by updateTraveledRouteLine()
         // won't have to wait for the result. The updateTraveledRouteLine method is much
         // more time sensitive.
-        if (routeLineOptions.styleInactiveRouteLegsIndependently) {
-            when (routeLineOptions.vanishingRouteLine) {
-                // If the styleInactiveRouteLegsIndependently feature is enabled but the
-                // vanishingRouteLine feature is not enabled then side effects are generated and
-                // need to be rendered.
-                null -> highlightActiveLeg(routeProgress, consumer)
-                else -> {
-                    ifNonNull(routeProgress.currentLegProgress) { routeLegProgress ->
-                        if (routeLegProgress.legIndex > activeLegIndex) {
-                            jobControl.scope.launch(Dispatchers.Main) {
-                                mutex.withLock {
-                                    alternativelyStyleSegmentsNotInLeg(
-                                        routeLegProgress.legIndex,
-                                        routeLineExpressionData
-                                    )
-                                    activeLegIndex = routeLegProgress.legIndex
+        when {
+            routeLineOptions.styleInactiveRouteLegsIndependently -> {
+                when (routeLineOptions.vanishingRouteLine) {
+                    // If the styleInactiveRouteLegsIndependently feature is enabled but the
+                    // vanishingRouteLine feature is not enabled then side effects are generated and
+                    // need to be rendered.
+                    null -> highlightActiveLeg(
+                        routeProgress,
+                        routeLineMaskingLayerDynamicData,
+                        consumer
+                    )
+                    else -> {
+                        ifNonNull(routeProgress.currentLegProgress) { routeLegProgress ->
+                            if (routeLegProgress.legIndex > activeLegIndex) {
```

Comment:
This check is always false since you are assigning `activeLegIndex = routeProgress.currentLegProgress?.legIndex` above.